subprojects/core/src/main/java/org/gradle/internal/scan/config/BuildScanPluginAutoApply.java

```diff
@@ -0,0 +1,111 @@
+import org.gradle.plugin.management.internal.DefaultPluginRequests;
+import org.gradle.plugin.management.internal.PluginRequestInternal;
+import org.gradle.plugin.management.internal.PluginRequests;
+import org.gradle.plugin.management.internal.PluginRequestsTransformer;
+import org.gradle.plugin.use.PluginId;
+import org.gradle.plugin.use.internal.DefaultPluginId;
+import org.gradle.util.CollectionUtils;
+
+import java.util.List;
+
+/**
+ * Automatically adds a plugin request for the build scan plugin when {@code --scan} is detected.
+ *
+ * <ol>
+ * <li>Plugin request is only added for the root project.</li>
+ * <li>No request is added if the project has already requested the build scan plugin, via {@code plugins} or {@code buildscript}.</li>
+ * <li>The plugin request is inserted before any other plugin requests.</li>
+ * <li>A fixed version of the build scan plugin is requested.</li>
+ * </ol>
+ *
+ * @since 4.3
+ */
+public class BuildScanPluginAutoApply implements PluginRequestsTransformer {
+    public static final String BUILD_SCAN_PLUGIN_AUTO_APPLY_VERSION = "1.10";
+    public static final PluginId BUILD_SCAN_PLUGIN_ID = DefaultPluginId.of("com.gradle.build-scan");
+    private static final String BUILD_SCAN_PLUGIN_GROUP = "com.gradle";
+    private static final String BUILD_SCAN_PLUGIN_NAME = "build-scan-plugin";
+
+    private final StartParameter startParameter;
+
+    public BuildScanPluginAutoApply(StartParameter startParameter) {
+        this.startParameter = startParameter;
+    }
+
+    @Override
+    public PluginRequests transformPluginRequests(PluginRequests requests, Object pluginTarget) {
+        if (!startParameter.isBuildScan()) {
+            return requests;
+        }
+
+        if (!isRootProject(pluginTarget)) {
+            return requests;
+        }
+
+        if (isPluginAlreadyRequested(requests, (Project) pluginTarget)) {
+            return requests;
+        }
+
+        DefaultPluginRequest buildScanPluginRequest = new DefaultPluginRequest(BUILD_SCAN_PLUGIN_ID, BUILD_SCAN_PLUGIN_AUTO_APPLY_VERSION, true, 0, "auto-apply", null);
+        return prependRequest(buildScanPluginRequest, requests);
+    }
+
+    private PluginRequests prependRequest(PluginRequestInternal buildScanPluginRequest, PluginRequests requests) {
+        List<PluginRequestInternal> copyRequests = Lists.newArrayList();
+        copyRequests.add(buildScanPluginRequest);
+        CollectionUtils.addAll(copyRequests, requests);
+        return new DefaultPluginRequests(copyRequests);
+    }
+
+    private boolean isRootProject(Object pluginTarget) {
+        if (!(pluginTarget instanceof Project)) {
+            return false;
+        }
+
+        Project project = (Project) pluginTarget;
+        return project.getGradle().getParent() == null
+            && project == project.getRootProject();
+    }
+
+    private static boolean isPluginAlreadyRequested(PluginRequests requests, Project project) {
```

Comment:
❌ This is missing the case where the scan plugin was added by an init script - in that case it is not in the requests and not in the `classpath` configuration, but it will be in `project.getPlugins`.

Reply:
done